### PR TITLE
[VecOps] Allow implicit conversions in Where

### PR DESCRIPTION
**Pull request by eguiraud**
Before this commit, something like this would fail:

    Where(arr1 > 2, 10, ROOT::RVecF{arr1.size(), 1.f})

Because 10 is not exactly `float`.
We now allow the scalar arguments to be of a different type from the
value_type of the RVec arguments if they can be implicitly coverted
to the value_type.

### CODE REVIEW

File: math/vecops/inc/ROOT/RVec.hxx [hunk 1]
```diff
@@ -2530,8 +2530,8 @@ RVec<T> Where(const RVec<int>& c, const RVec<T>& v1, T v2)
 /// if_c_v1_else_v2
 /// // (ROOT::VecOps::RVec<double>) { 1.0000000, 4.0000000, 4.0000000 }
 /// ~~~
-template <typename T>
-RVec<T> Where(const RVec<int>& c, T v1, const RVec<T>& v2)
+template <typename T, typename U, typename = std::enable_if<std::is_convertible<T, U>::value, void>>
```
Comment (Axel-Naumann) [inline math/vecops/inc/ROOT/RVec.hxx:2533]:
```suggestion
template <typename T, typename U, typename = typename std::enable_if<std::is_convertible<T, U>::value>::type>
```
or what am I missing? (And yes we need concepts.)

**Review by Axel-Naumann — APPROVED** [review]
I find this a bit dangerous. These functions take a "boolean" vector (the `RVecI`) which picks from `v1` or `v2`. That's symmetrical, and before, the return type was obvious because it was the same for `v1` and `v2`. Now, the return type is determined by `v1`. If `v2` converts to `v1`: good. If `v1` converts to `v2`: bad.

*If* you really want say `Where(c, derived, base)` then you probably want symmetry, whether `derived` is the first or the second argument? So the return type should be `RVec<typename common_type<T, U>::type>`?

File: math/vecops/inc/ROOT/RVec.hxx (hunk 1 above)
Comment (eguiraud) [reply math/vecops/inc/ROOT/RVec.hxx:2533]: ah wow, of course I meant `std::enable_if_t`, thanks!

**Comment by eguiraud** [comment]
@Axel-Naumann fair enough, I'll just fix the case I mention in my first comment which is not ambiguous (one RVec and one scalar, we pick the type of the RVec for the return type).

**Comment by Axel-Naumann** [comment]
Will both
```c++
Where(arr1 > 2, 10, ROOT::RVecF{arr1.size(), 1.f})
```
and
```c++
Where(arr1 > 2, ROOT::RVecF{arr1.size(), 1.f}, 10)
```
have the same return type?

**Comment by eguiraud** [comment]
Yes to the first question, I presume yes to the second (I expect vastly more common usecase to be that the user typed `10` instead of `10.f` not to cleverly coerce the return type, but because they couldn't bother/didn't think about it).

> Can we use (skipping const & etc) Where(RVec<int>, RVec<T>, typename RVec<T>::element_type) as signature?

Yes, and that's a better solution :+1: Will do

**Pull request merged**